### PR TITLE
Pick starting spots for empires with a minimum variety of planets

### PR DESCRIPTION
**Pull request by MichaelCourtney**

### CODE REVIEW

**Comment by MichaelCourtney** [comment]
https://freeorion.org/forum/viewtopic.php?t=12779

**Comment by Cjkjvfnby** [comment]
I'd recommend you install pre-commit hooks locally, it will be much easier that checking logs for linter reports. 

In this particular case hooks this sort of import for you, so you don't need to memorize the rules behind this.

https://github.com/freeorion/freeorion/blob/master/CONTRIBUTING.md#installing-pre-commit-hooks-optional

File: default/python/universe_generation/empires.py
Comment (Cjkjvfnby) [inline default/python/universe_generation/empires.py]: You could use a set to simplify these operations.

```python
a = [1, 2, 3]
b = [2, 3]
set(a).difference(b)
{1}
```

Probably you could return set from this function, but if not, keep the list.

File: default/python/universe_generation/empires.py
Comment (Cjkjvfnby) [inline default/python/universe_generation/empires.py]: This means that we have only one planet of each type, maybe we could avoid this. because this limits the max number of planets by a number of types.  And if we have very few types acceptable, we will hit the limit more often.

We could just sort `acceptable_planet_types` each time we do the query.

File: default/python/universe_generation/natives.py
Comment (Cjkjvfnby) [inline default/python/universe_generation/natives.py]: We could use the Python built-in feature to deal with that.  It still allows duplications but will make them controllable.

I use a generator, to avoid extra code that accumulates the result. 


```python
import random


def _build_wights(items: list[str], number_of_items_to_select: int):
    return [number_of_items_to_select for _ in items]


def select_recent(items: list[str], number_of_items_to_select, reduce_factor=1):
    """
    Select the number_of_items_to_select items, each time reducing the probability to select the same item.

    reduce_factor is define the probability to choose the unique item.
    1 is maximum duplication are allowed, >= len(items) no duplication are allowed.

    >>> from collections import Counter
    >>> Counter(select_recent(["a", "b", "c"], 3, reduce_factor=3)) == Counter({"a": 1, "b": 1, "c": 1})

    If factor is > 1 and number_of_items_to_select is bigger than number of items, duplication are still possible.

    For example, this code will produce exactly 2 sets of unique items

     >>> from collections import Counter
    >>> Counter(select_recent(["a", "b", "c"], 6, reduce_factor=3)) == Counter({"a": 2, "b": 2, "c": 2})

    """
    weights = _build_wights(items, number_of_items_to_select)

    for _ in range(number_of_items_to_select):
        try:
            choice = random.choices(items, weights=weights, k=1)[0]
        except ValueError:
            # all weights are zero
            weights = _build_wights(items, number_of_items_to_select)
            choice = random.choices(items, weights=weights, k=1)[0]
        yield choice
        weights[items.index(choice)] = max(weights[items.index(choice)] - reduce_factor, 0)


print(list(select_recent(["a", "b", "c"], 3, reduce_factor=3)))
```

**Comment by MichaelCourtney** [comment]
> I'd recommend you install pre-commit hooks locally, it will be much easier that checking logs for linter reports.
> 
> In this particular case hooks this sort of import for you, so you don't need to memorize the rules behind this.
> 
> https://github.com/freeorion/freeorion/blob/master/CONTRIBUTING.md#installing-pre-commit-hooks-optional

I'm having trouble getting this to work under windows, and all the instructions I find seem to assume I'm on linux. pip install pre-commit ran, and the package seems to be on my computer but then when I run "pre-commit install" it doesn't seem to find it. maybe a path issue? anyway I'll try and get it working

**Comment by MichaelCourtney** [comment]
ok it was a path issue, seems you have to add this manually on windows

File: default/python/universe_generation/empires.py
Comment (Vezzra) [reply default/python/universe_generation/empires.py]: The loop iterates over the planet types, so each type gets checked once, so there cannot be any duplicates in the resulting list. Hence it should make no difference if a list or set is used, unless I'm missing something...?

Comment (Cjkjvfnby) [reply default/python/universe_generation/empires.py]: This loop and if statement could be replaced by a single `difference` call.  This will be much more readable code.

Comment (MichaelCourtney) [reply default/python/universe_generation/empires.py]: https://github.com/freeorion/freeorion/compare/2e7e0c095411c8deb2d879a9a7d46685816b0ade..cbc21cc05bd8189847d7ffaaf3f5e30793e10537

File: default/python/universe_generation/empires.py
Comment (MichaelCourtney) [reply default/python/universe_generation/empires.py]: Added a bool to allow for both repeating the planet type and not 
https://github.com/freeorion/freeorion/compare/357e793954c62ee320eeaae432262a2db54404b7..436b15946522ba62df855297b7526605e90c43bd

File: default/python/universe_generation/natives.py
Comment (MichaelCourtney) [reply default/python/universe_generation/natives.py]: This one I plan to rewrite so it knows upfront how many natives its placing, so it can work out how many repeats are necessary

**Comment by MichaelCourtney** [comment]
Dropped changes to natives (will put these in a future PR), cleaned up scripting in line with Cjkjvfnby's suggestions(hopefully), added options to ensure habitable planet near starting position.

**Comment by Cjkjvfnby** [comment]
@MichaelCourtney do you need any help with this one?

**Comment by MichaelCourtney** [comment]
> @MichaelCourtney do you need any help with this one?

Should be all good to go, pending approval. Works fine on my end.

File: default/python/universe_generation/empires.py
Comment (Cjkjvfnby) [inline default/python/universe_generation/empires.py:92]: You could also simplify this one:

Keep in mind that this is changing the return type to set as well.  I don't think that list has since for this data, but if you want to keep it for compatibility just replace `{}` with `[]`

Also, I recommend to keep the function, even if it's a one-line.  Expression is a bit complex, so it's good to have a name for it.
```suggestion
    return {fo.planet_get_type(planet) for system in systems for planet in fo.sys_get_planets(system)}
```

Maybe you will be able to simplify count_planet_types_in_systems after that.

File: default/python/universe_generation/empires.py
Comment (Cjkjvfnby) [inline default/python/universe_generation/empires.py:579]: That means that some users will get `planetEnvironment.good` and others `planetEnvironment.adequate`. This does not look fair.  Should it be the same planet type for everyone?

File: default/python/universe_generation/empires.py
Comment (Cjkjvfnby) [inline default/python/universe_generation/empires.py:583]: It's optional.

You could try to convert all new functions to sets, in that case, you won't need this `set(planet_types_in_vicinity)`

**Review by Cjkjvfnby — COMMENTED** [review]
Left a couple of cosmetic comments.

@Vezzra do you want to take a look?  Otherwise, we can merge changes.

File: default/python/universe_generation/empires.py
Comment (MichaelCourtney) [reply default/python/universe_generation/empires.py:579]: fairness is hard to guarantee without it always being an average sized good on a yellow star (or GG for Sly). The narrow tolerance species don't have an adequate type for instance. I didn't have a great answer to the fairness question, so I just implemented a simple version and defaulted the option to off. I'm not convinced it matters for single player and multiplayer probably needs something trickier to be fair.

**Comment by Vezzra** [comment]
> @Vezzra do you want to take a look? Otherwise, we can merge changes.

You seem to have taken a quite thorough look at this already, that's good enough for me. Between the two of us you're definitely the more competent Python coder. :wink:

And unfortunately I don't really have the time to look at this thoroughly. I have taken a quick look, and would have worked even more with sets instead of lists (or dicts if counting how much of each planet type is or isn't present in the "near vicinity" was desired).

That said, if you think the code is ok so far, and it works as intended, then I'd say, LGTM.

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**
